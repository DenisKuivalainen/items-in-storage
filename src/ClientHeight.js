const { useState, useLayoutEffect } = require("react");

const useClientHeight = (sub) => {
    const [size, setSize] = useState(0);
    const updateSize = () => setSize(document.documentElement.clientHeight);

    // Get window size
    useLayoutEffect(() => {
        window.addEventListener('resize', updateSize);
        updateSize();
        return () => window.removeEventListener('resize', updateSize);
    }, []);

    // Get client size
    sub = !!sub ? sub : 0;

    return size - sub;
}

module.exports = { useClientHeight };