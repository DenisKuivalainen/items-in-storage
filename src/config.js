const { __ } = require("ramda");
const { viewOnPath } = require("ramda-godlike");
const configFile = require("../config.json");

const config = (...args) => viewOnPath(args, configFile);

module.exports = { config };