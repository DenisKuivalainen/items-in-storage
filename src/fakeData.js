const { times, identity, map, toString, splitEvery, join, compose, curry } = require("ramda");

const getValue = curry((n, l) => {
    let letters = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"];

    return compose(
        x => l + "-" + x,
        join(""),
        map(x => letters[parseInt(x)]),
        splitEvery(1),
        toString
    )(n)

});

const getObject = (val) => {
    const gv = getValue(val);

    return { "id": val, "type": gv("tp"), "name": gv("nm"), "color": gv("cl"), "price": ((val + 254) * 3) + "$", "manufacturer": gv("mn"), "availability": gv("av") }
}

const data = (n) => map(getObject, times(identity, n));

module.exports = { data };