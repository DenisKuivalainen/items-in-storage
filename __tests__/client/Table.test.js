import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from "@testing-library/react";
import TableElement from "../../src/client/Table";
import { data } from "../../src/fakeData";

describe("test TableElement ", () => {
    const Table = ({n}) => TableElement(data(n), 500);

    test("sanpshot with values length of 4", () => {
        const tree = renderer.create(<Table n={4}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("forward and backward buttons functionality", () => {
        const { getByTitle } = render(<Table n={24}/>);
        const towardBtn = getByTitle("Next page");
        const backBtn = getByTitle("Previous page");
        const getCellValue = () => screen.getAllByRole('cell')[0].textContent;

        // id of element n will be lower than id of element n + m (n >= 0 && m > 0)
        // So, if I want to check element on page 2 (we have m elements per page), 
        // the first element of the page will be n + m, while on the page 1 it will be n

        let first = getCellValue();
        
        // Go 2 pages forward
        fireEvent.click(towardBtn);
        fireEvent.click(towardBtn);
        let third = getCellValue();

        // Go 1 page backward
        fireEvent.click(backBtn);
        let second = getCellValue();

        expect(first).toEqual("0");
        expect(second).toEqual("10");
        expect(third).toEqual("20");
    });
})