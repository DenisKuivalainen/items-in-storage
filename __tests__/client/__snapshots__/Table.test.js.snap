// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test TableElement  sanpshot with values length of 4 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="MuiTableContainer-root"
    style={
      Object {
        "maxHeight": 500,
      }
    }
  >
    <table
      aria-label="sticky table"
      className="MuiTable-root MuiTable-stickyHeader"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 150,
              }
            }
          >
            ID
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 200,
              }
            }
          >
            Name
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 150,
              }
            }
          >
            Manufacturer
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            Price
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            Color
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            Availability
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-hover"
          role="checkbox"
          tabIndex={-1}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 150,
              }
            }
          >
            0
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 200,
              }
            }
          >
            nm-a
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 150,
              }
            }
          >
            mn-a
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            762$
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            cl-a
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            av-a
          </td>
        </tr>
        <tr
          className="MuiTableRow-root MuiTableRow-hover"
          role="checkbox"
          tabIndex={-1}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 150,
              }
            }
          >
            1
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 200,
              }
            }
          >
            nm-b
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 150,
              }
            }
          >
            mn-b
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            765$
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            cl-b
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            av-b
          </td>
        </tr>
        <tr
          className="MuiTableRow-root MuiTableRow-hover"
          role="checkbox"
          tabIndex={-1}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 150,
              }
            }
          >
            2
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 200,
              }
            }
          >
            nm-c
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 150,
              }
            }
          >
            mn-c
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            768$
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            cl-c
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            av-c
          </td>
        </tr>
        <tr
          className="MuiTableRow-root MuiTableRow-hover"
          role="checkbox"
          tabIndex={-1}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 150,
              }
            }
          >
            3
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 200,
              }
            }
          >
            nm-d
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 150,
              }
            }
          >
            mn-d
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            771$
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            cl-d
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            av-d
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        className="MuiTablePagination-spacer"
      />
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        className="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          10
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value={10}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        1-4 of 4
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Previous page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Next page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
